pil_analyzer/src/pil_analyzer.rs

```diff
@@ -865,4 +865,74 @@ namespace N(65536);
         let formatted = process_pil_file_contents::<GoldilocksField>(input).to_string();
         assert_eq!(formatted, input);
     }
+
+    #[test]
+    fn symbolic_functions() {
+        let input = r#"namespace N(16);
+    let last_row = 15;
+    let ISLAST = |i| match i { last_row => 1, _ => 0 };
+    let x;
+    let y;
+    let constrain_equal_expr = |A, B| A - B;
+    let on_regular_row = |cond| (1 - ISLAST) * cond;
+    on_regular_row(constrain_equal_expr(x', y)) = 0;
+    on_regular_row(constrain_equal_expr(y', x + y)) = 0;
+    "#;
+        let expected = r#"constant last_row = 15;
+namespace N(16);
+    col fixed ISLAST(i) { match i { last_row => 1, _ => 0, } };
+    col witness x;
+    col witness y;
+    let constrain_equal_expr = |A, B| (A - B);
+    col fixed on_regular_row(cond) { ((1 - N.ISLAST) * cond) };
+    ((1 - N.ISLAST) * (N.x' - N.y)) = 0;
+    ((1 - N.ISLAST) * (N.y' - (N.x + N.y))) = 0;
+"#;
+        let formatted = process_pil_file_contents::<GoldilocksField>(input).to_string();
+        assert_eq!(formatted, expected);
+    }
+
+    #[test]
+    fn next_op_on_param() {
+        let input = r#"namespace N(16);
+    let last_row = 15;
+    let ISLAST = |i| match i { last_row => 1, _ => 0 };
+    let x;
+    let y;
+    let next_is_seven = |t| t' - 7;
+    next_is_seven(y) = 0;
+    "#;
+        let expected = r#"constant last_row = 15;
+namespace N(16);
+    col fixed ISLAST(i) { match i { last_row => 1, _ => 0, } };
+    col witness x;
+    col witness y;
+    col fixed next_is_seven(t) { (t' - 7) };
+    (N.y' - 7) = 0;
+"#;
+        let formatted = process_pil_file_contents::<GoldilocksField>(input).to_string();
+        assert_eq!(formatted, expected);
+    }
+
+    #[test]
+    fn fixed_concrete_and_symbolic() {
+        let input = r#"namespace N(16);
+    let last_row = 15;
+    let ISLAST = |i| match i { last_row => 1, _ => 0, };
+    let x;
+    let y;
+    y - ISLAST(3) = 0;
+    x - ISLAST = 0;
```

Comment:
It makes sense, but I find it confusing that ISLAST(3) evaluates at compile-time and ISLAST compiles to a runtime column, even though both look kind of the same syntactically.

Reply:
Once we have a distinction by types, we could change it so that you can never evaluate fixed columns in an algebraic expression context, but that could also create more problems than it solves.

Another option would be to require explicit syntax to turn a symbol into an actual column reference. But then it would look quite ugly:

```
ref(y) - ISLAST(3) = 0
ref(x) - ref(ISLAST) = 0
```

I think that in the end, people will just not notice this difference between ISLAST and ISLAST(3) and find it quite natural.

Reply:
yea maybe